src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -1043,8 +1103,20 @@ public int doUpdate(byte[] in, int inOfs, int inLen, byte[] out,
             }
 
             // Encrypt the remaining blocks inside of 'in'
+            if (inLen >= PARALLEL_LEN) {
+                int r = GaloisCounterMode.implGCMCrypt(in, inOfs, inLen, out,
+                    outOfs, out, outOfs, gctr, ghash);
+                len += r;
+                inOfs += r;
+                inLen -= r;
+                outOfs += r;
+            }
+
             if (inLen >= blockSize) {
-                len += gctrghash.update(in, inOfs, inLen, out, outOfs);
+                int r = op.update(in, inOfs, inLen, out, outOfs);
+                len += r;
+                inOfs += r;
+                inLen -= r;
             }
```

Comment:
Have you considered move the "if (inLen >= PARALLEL_LEN) block" into EncryptOp.update() impl (just like the Encrypt.doFinal() impl) ? Even though not all op.update() calls process large data, but it'd reduce code duplication and ensures that all large data processed by EncryptOp.update() calls would call the intrinsified method.

Reply:
There are cases where inLen is known to be smaller than PARALLEL_LEN and is a waste of a check, such as merging with the ibuffer to create one block.  Also moving it into EncryptOp would always mean an additional check and maybe an unnecessary jump to another method.

I did that for doFinal, because gctr/ghash.doFinal() needs to was no extra checks.

Reply:
Alright, as long as you did consider that.